### ISSUE TITLE
Add Contributor model

### ISSUE THREAD

**Issue by joshsmith** [code-corps/deprecated-code-corps-rails-api#12]
We need a `Contributor` model that relates a `User` to a `Project`, with some metadata like their status in the project (admin, etc). We should look to GitHub for a good set of statuses.

**Comment by begedin** [comment]
Github differentiates between a repository owned by a user or one owned by an organization.
- [A Contributor on a user owned repository can be either an owner or collaborator](https://help.github.com/articles/permission-levels-for-a-user-account-repository/)
- For an organization owned repository, access levels can be granted either to teams, or individual users. It's a pretty complex system.
  - [Each team or individual user can have read, write or admin permissions on a project](https://help.github.com/articles/repository-permission-levels-for-an-organization)
  - Each user can be an owner or a member of an organization
  - An organization owner has additional permissions (more than admin) on every project
  - There is also an "outside collaborator" status a non-member of an organization can have on projects belonging to an organization

**Comment by begedin** [comment]
@JoshSmith This is blocking a large number of our tasks, so we need to figure it out. My main source of confusion is how to take organization-owned projects into account.

The way I see it

If user owns a project
- There's a `Contributor` record for that user with role `:owner`
- Other users can request collaboration on the project
  - If the owner approves, a `Contributor` record is created with role `:collaborator`
  - Owner can later change role levels for these `Contributor` records, though I'm not sure which other level's we'd need in this case.

If organization owns a project
- Users can request collaboration, organization owner or admin can approve them, creating `Contributor` records
- Organization assigns a team to the project. What happens?
  - Do we create `Contributor` records for each team member? 
  - Can the roles of these records be altered later, regardless of team status?
  - Can the team be unassigned from the project? What happens to the records then?
  - In any case, we could use an additonal `Contributor` role - `:admin`. This one can be used to allow some user to approve collaboration requests by other users.

Overall, i believe roles/statuses for `Contributor` records should be
- `:owner` - highest role, can make other's into admins
- `:admin` - can approve requests for colaboration
- `:collaborator` - can contribute to the project and that's it
- `:pending` - a newly created request for collaboration, not approved by owner or admin yet

**Comment by joshsmith** [comment]
I think I agree with everything you said except for the idea of there being `Contributor` records for the team members. I think a `Team` also has roles, which are `admin` and `collaborator`, which varies from project to project.

When you verify whether someone has the ability to do something, like post a particular type of post, we'd first check their direct collaborator status, then if they're on one of that projects teams.

It's possible that `CanCanCan` doesn't scale well for this. Maybe we should consider [Pundit](https://github.com/elabs/pundit)?

What do you think?

**Comment by begedin** [comment]
I'm not familiar with Pundit, but based on what I'm seeing, it does seem more fitting, since we use different sets of roles in different scenarios (as in, team management, project management, organizational management, handling project contributions, etc.).

**Comment by begedin** [comment]
To add, I think the first step in that case should be to move the existing model to Pundit.

**Comment by joshsmith** [comment]
Okay, I agree. Does this seem like the logical thing to do? You agree with what I said earlier?

**Comment by joshsmith** [comment]
@begedin don't want to block you on so much so is this unblocking?

**Comment by begedin** [comment]
Yeah. I was trying to fit our model into CanCan, but what you propose makes more sense, especially with something like pundit. 

You can consider things to be unblocked now. Should I create a separate issue for the existing roles from moving from CanCan to Pundit, or do I do it as part of this one?

**Comment by joshsmith** [comment]
I think a separate task.

**Comment by begedin** [comment]
Blocked by #67
